contracts/TokenDistributor.sol

```diff
@@ -0,0 +1,212 @@
+    /// @inheritdoc ITokenDistributor
+    uint[1000000000000000] public tokensPerEpoch;
+
+    /// @notice the week offset in seconds
+    uint internal offset;
+
+    /// @notice max amount of days the epoch can be offset by
+    uint internal constant MAX_OFFSET_DAYS = 6;
+
+    /// @notice weeks in a year
+    uint internal constant WEEKS_IN_YEAR = 52;
+
+    /// @notice constructs the TokenDistributor contract
+    /// and sets startTime
+    /// @param _kwenta: address of the kwenta contract
+    /// @param _stakingRewardsV2: address of the stakingRewardsV2 contract
+    /// @param _rewardEscrowV2: address of the rewardEscrowV2 contract
+    constructor(
+        address _kwenta,
+        address _stakingRewardsV2,
+        address _rewardEscrowV2,
+        uint daysToOffsetBy
+    ) {
+        kwenta = IKwenta(_kwenta);
+        stakingRewardsV2 = StakingRewardsV2(_stakingRewardsV2);
+        rewardEscrowV2 = RewardEscrowV2(_rewardEscrowV2);
+
+        /// @notice custom start day (startTime + daysToOffsetBy)
+        if (daysToOffsetBy > MAX_OFFSET_DAYS) {
+            revert OffsetTooBig();
+        }
+        offset = daysToOffsetBy * 1 days;
+        uint startOfThisWeek = _startOfWeek(block.timestamp);
+        startTime = startOfThisWeek;
+        lastCheckpoint = startOfThisWeek;
+    }
+
+    /// @inheritdoc ITokenDistributor
+    function checkpointToken() public override {
+        uint tokenBalance = kwenta.balanceOf(address(this));
+        uint toDistribute = tokenBalance - lastTokenBalance;
+        lastTokenBalance = tokenBalance;
+
+        uint previousCheckpoint = lastCheckpoint;
+        uint sinceLast = block.timestamp - previousCheckpoint;
+        lastCheckpoint = block.timestamp;
+        uint thisWeek = _startOfWeek(previousCheckpoint);
+        uint nextWeek = 0;
+
+        /// @dev Loop for potential missed weeks
+        /// iterates until caught up, unlikely to go to 52 weeks
+        for (uint i = 0; i < WEEKS_IN_YEAR; i++) {
+            nextWeek = thisWeek + 1 weeks;
+
+            if (block.timestamp < nextWeek) {
+                /// @dev if in the current week
+                if (sinceLast == 0 && block.timestamp == previousCheckpoint) {
+                    /// @dev If no time change since last checkpoint just add new tokens
+                    /// that may have been deposited (same block)
+                    tokensPerEpoch[thisWeek] += toDistribute;
+                } else {
+                    /// @dev In the event that toDistribute contains tokens
+                    /// for multiple weeks we take the remaining portion
+                    tokensPerEpoch[thisWeek] +=
+                        (toDistribute *
+                            (block.timestamp - previousCheckpoint)) /
+                        sinceLast;
+                }
+                break;
+            } else {
+                /// @dev If passed weeks missed
+                if (sinceLast == 0 && nextWeek == previousCheckpoint) {
```

Comment:
**Issue:** If `sinceLast == 0`, then surely it is not possible to be in this else block?
Because if that is the case, then the last checkpoint was in the very same second as now. And therefore `block.timestamp < nextWeek` cannot be true?

As if `block.timestamp == previousCheckpoint`, then `thisWeek` must be at the very smallest `block.timestamp - 604799`, and `nextWeek = block.timestamp + 604800`, therefore if `sinceLast = 0`, then `nextWeek` must be at least 1 greater than `block.timestamp`.

This means that `block.timestamp < nextWeek` will never evaluate as true if `sinceLast == 0` and hence this if block is unnecessary.

Definitely double check my logic, but pretty sure about this.